### PR TITLE
Address edge scenarios with JsonSerializer's property visibility

### PR DESCRIPTION
**Pull request by layomia**
Contributes to fixing https://github.com/dotnet/runtime/issues/37640. Will leave the issue open until the fix is confirmed in app compat. cc @DotNetAppCompatFeiWang

https://github.com/dotnet/runtime/pull/36936 made a change so that the serializer ignores virtual properties when a derived property that hid them is ignored (with `[JsonIgnore]`) **_and_** there's a JSON property name collision.

The failing scenario in https://github.com/dotnet/runtime/issues/37640 makes it clear that we need to ignore virtual properties when a derived property that hid them is ignored, whether or not there's a property name collision. This is to avoid serializing content that the caller explicitly marked to be ignored.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs [hunk 1]
```diff
@@ -261,8 +261,8 @@ public static void VeryLargeAmountOfEnumsToSerialize()
             // Use multiple threads to perhaps go over the soft limit of 64, but not by more than a couple.
             Parallel.For(0, 8, i => JsonSerializer.Serialize((MyEnum)(46 + i), options));
 
-            // Write the remaining enum values. We should not store any more values in
-            // the cache. If we do, we may throw OutOfMemoryException on some machines.
+            // Write the remaining enum values. The cache is capped to avoid
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/EnumConverterTests.cs:264]: Following up on https://github.com/dotnet/runtime/pull/37710#discussion_r438372246.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 2]
```diff
@@ -138,10 +142,7 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                                             !jsonPropertyInfo.IsIgnored &&
                                             // Is the current property hidden by the previously cached property
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:143]: Are these comments out-of-date now regarding "hidden by the previously..."?

**Comment by steveharter** [comment]
Are the current semantics in this PR consistent with Newtonsoft?

**Comment by layomia** [comment]
> Are the current semantics in this PR consistent with Newtonsoft?

With the latest commit, we are largely aligned with Newtonsoft.Json behavior.

- If an override of a virtual property is `[JsonIgnore]d`, then the property is ignored as well.

https://github.com/dotnet/runtime/blob/766572fd4cc8fc2bae7d1cde857aafe43473c71b/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs#L386-L387

- If a property that hides a virtual or non-virtual property is `[JsonIgnore]d`, and there's no property name collision, then the hidden property is included for (de)serialization.

https://github.com/dotnet/runtime/blob/766572fd4cc8fc2bae7d1cde857aafe43473c71b/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs#L395-L396

- If a property that hides a virtual or non-virtual property is `[JsonIgnore]d`, and there's a property name collision between a non-ignored property and the hidden property, then the hidden property is ignored, while the non-ignored property is included for (de)serialization

https://github.com/dotnet/runtime/blob/766572fd4cc8fc2bae7d1cde857aafe43473c71b/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs#L410-L411

wrt. how we differ - in Newtonsoft.Json, property name conflicts at different type-hierarchy levels that are not caused by deriving or the new keyword are allowed. So, properties on more derived types win. In System.Text.Json, property name conflicts are only valid when caused by deriving or the new keyword.

https://github.com/dotnet/runtime/blob/766572fd4cc8fc2bae7d1cde857aafe43473c71b/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs#L427

**Comment by steveharter** [comment]
> wrt. how we differ - in Newtonsoft.Json, property name conflicts at different type-hierarchy levels that are not caused by deriving or the new keyword are allowed. So, properties on more derived types win. In System.Text.Json, property name conflicts are only valid when caused by deriving or the new keyword.

This means all CLR naming "conflicts" are OK, but naming conflicts through `[JsonPropertyName]` are not?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:7]: Is this necessary (I didn't see any use of nullability attributes)?

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:7]: No - will remove. I used a nullability attribute in a helper in an earlier iteration of this change.

**Comment by layomia** [comment]
> This means all CLR naming "conflicts" are OK, but naming conflicts through [JsonPropertyName] are not?

Yes. `[JsonPropertyName]`  and also `JsonNamingPolicy`.

**Review by steveharter — APPROVED** [review]

**Comment by DotNetAppCompatFeiWang** [comment]
cc @dotnet-actwx-bot

**Pull request merged**